torch/_functorch/autograd_function.py

```diff
@@ -720,33 +720,68 @@ def __call__(self, fwd, bwd, *fwd_args, **fwd_kwargs):
         saved_values = None
         args_tensor_mask = fwd_kwargs["args_tensor_mask"]
         non_differentiable_idx = fwd_kwargs["non_differentiable_idx"]
+        is_setup_ctx_defined = fwd_kwargs["is_setup_ctx_defined"]
+        original_generate_vmap_rule = fwd_kwargs["generate_vmap_rule"]
         length_of_tensor_args = sum(args_tensor_mask)
         # Filter out the original tensor args from fwd_args,
         # lifted freevars should not be args of ApplyTemplate.apply
         # since we don't need to calculate the gradients of them.
         new_fwd_args = fwd_args[:length_of_tensor_args]
 
-        class ApplyTemplate(torch.autograd.Function):
-            @staticmethod
-            def forward(ctx, *args):
-                nonlocal saved_values
-                output, saved_values = fwd(None, *fwd_args)
+        # If users call ctx.mark_non_differentiable() in the original fwd function.
+        def _ctx_mark_non_differentiable(ctx, output, non_differentiable_idx):
+            if len(non_differentiable_idx) > 0:
+                non_differentiable_output = []
+                for i, x in enumerate(output):
+                    if i in non_differentiable_idx:
+                        non_differentiable_output.append(x)
+                ctx.mark_non_differentiable(*non_differentiable_output)
 
-                # If users call ctx.mark_non_differentiable() in the original fwd function.
-                if len(non_differentiable_idx) > 0:
-                    non_differentiable_output = []
-                    for i, x in enumerate(output):
-                        if i in non_differentiable_idx:
-                            non_differentiable_output.append(x)
-                    ctx.mark_non_differentiable(*non_differentiable_output)
+        if is_setup_ctx_defined:
 
-                return output
+            class ApplyTemplate(torch.autograd.Function):
+                nonlocal original_generate_vmap_rule
+                generate_vmap_rule = original_generate_vmap_rule
 
-            @staticmethod
-            def backward(ctx, *grad):
-                return bwd(None, *grad, *saved_values)
+                @staticmethod
+                def forward(*args):
+                    nonlocal saved_values
+                    output, saved_values = fwd(None, *fwd_args)
+                    return output
 
-        return ApplyTemplate.apply(*new_fwd_args)
+                @staticmethod
+                def setup_context(ctx, inputs, output):
+                    nonlocal saved_values
+                    ctx.save_for_backward(*saved_values)
```

Comment:
Actually this is not 100% correct, since ```save_for_backward``` only supports tensor, but ```saved_values``` can be anything. In the non-transform mode, we pass these saved values via arguments directly, not sure how transform mode expect these values pass into backwad.

Reply:
Yeah. We can probably just separate them out based on if they're a tensor or not?

Can saved_values ever be a List[Tensor]? (I'm thinking No). If not then this isn't too bad